src/librustc_mir/interpret/eval_context.rs

```diff
@@ -121,31 +122,113 @@ impl<'mir, 'tcx: 'mir> PartialEq for Frame<'mir, 'tcx> {
     }
 }
 
-impl<'mir, 'tcx: 'mir> Hash for Frame<'mir, 'tcx> {
-    fn hash<H: Hasher>(&self, state: &mut H) {
-        let Frame {
-            mir: _,
-            instance,
-            span: _,
-            return_to_block,
-            return_place,
-            locals,
-            block,
-            stmt,
-        } = self;
+/// The virtual machine state during const-evaluation at a given point in time.
+#[derive(Eq, PartialEq)]
+struct EvalSnapshot<'a, 'mir, 'tcx: 'a + 'mir, M: Machine<'mir, 'tcx>>
+    (M, Vec<Frame<'mir, 'tcx>>, Memory<'a, 'mir, 'tcx, M>);
 
-        instance.hash(state);
-        return_to_block.hash(state);
-        return_place.hash(state);
-        locals.hash(state);
-        block.hash(state);
-        stmt.hash(state);
+impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> Hash for EvalSnapshot<'a, 'mir, 'tcx, M> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let EvalSnapshot(machine, stack, memory) = self;
+        EvalSnapshotRef(machine, stack, memory).hash(state);
     }
 }
 
-/// The virtual machine state during const-evaluation at a given point in time.
-type EvalSnapshot<'a, 'mir, 'tcx, M>
-    = (M, Vec<Frame<'mir, 'tcx>>, Memory<'a, 'mir, 'tcx, M>);
+#[derive(Eq, PartialEq)]
+struct EvalSnapshotRef<'a, 'b: 'a, 'mir: 'a, 'tcx: 'b + 'mir, M: Machine<'mir, 'tcx> + 'a>
+    (&'a M, &'a Vec<Frame<'mir, 'tcx>>, &'a Memory<'b, 'mir, 'tcx, M>);
+
+impl<'a, 'b, 'mir, 'tcx, M: Machine<'mir, 'tcx>> Hash for EvalSnapshotRef<'a, 'b, 'mir, 'tcx, M> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let EvalSnapshotRef(machine, stack, memory) = self;
+
+        machine.hash(state);
+
+        for frame in stack.iter() {
+            let Frame {
+                instance,
+                return_to_block,
+                return_place,
+                locals,
+                block,
+                stmt,
+                ..
+            } = frame;
+
+            instance.hash(state);
+            return_to_block.hash(state);
+            return_place.hash(state);
+            block.hash(state);
+            stmt.hash(state);
+
+            // Stores the hashes of all `Allocation`s that have already been visited
+            let mut allocs_seen = FxHashMap::<AllocId, u64>::default();
+
+            for local in locals.iter() {
+                // Skip dead locals
+                if let Some(value) = local {
+                    match value {
```

Comment:
I think you can just call `HashStable::hash_stable` on the value to obtain the same effect + simultaneously do the correct thing on `ByVal`. Have a look at the `HashStable` impl, it should look very much like your code.

Reply:
So I tried reusing `HashStable` for value, however [rustc panics](https://github.com/rust-lang/rust/blob/master/src/librustc/ich/impls_ty.rs#L415):

```
thread 'main' panicked at 'no value for AllocId', libcore/option.rs:989:5
```

I guess that means dangling references aren't expected when computing the stable hash, but on the other hand I don't see why that would make the hash _not stable_. Indeed, hashing the potentially empty result directly does seem to fix the issue. 

Does that makes sense?

Reply:
Yea... That's probably wrong. We should not ICE on

```rust
#![feature(const_let)]

const FOO: *const u32 = {
    let x = 42;
    &x
};

fn main() {
    let z = FOO;
}
```

and similar code. (If we fixed the ICE on the above code, we'd just end up with the ICE you're seeing) So we can just treat dangling pointers as not existing by not hashing them.

Reply:
Right, so if got what you're saying right, it's safe to fix the implementation of `HashStable` for `AllocId` such that it doesn't panic and then just reuse it to hash the `Frame::locals` here.

Reply:
yes